File: python/platforms/twitch.py
from __future__ import annotations
import os
from typing import Any
import requests
from core.models import InboundEvent, OutboundReply
from .base import PlatformAdapter
from .catalog import PLATFORM_CATALOG

class TwitchAdapter(PlatformAdapter):
    def __init__(self, token=None, client_id=None, broadcaster_id=None, sender_id=None, session=None):
        self.capabilities = PLATFORM_CATALOG["twitch"]; self.token = token or os.getenv("TWITCH_ACCESS_TOKEN"); self.client_id = client_id or os.getenv("TWITCH_CLIENT_ID"); self.broadcaster_id = broadcaster_id or os.getenv("TWITCH_BROADCASTER_ID"); self.sender_id = sender_id or os.getenv("TWITCH_SENDER_ID")
        if not all((self.token, self.client_id, self.broadcaster_id, self.sender_id)): raise ValueError("TWITCH_ACCESS_TOKEN, TWITCH_CLIENT_ID, TWITCH_BROADCASTER_ID and TWITCH_SENDER_ID are required")
        self.session = session or requests.Session()
    def parse_event(self, payload: Any, headers=None) -> InboundEvent:
        event = payload.get("event", payload)
        user_id = event.get("chatter_user_id") or event.get("user_id") or payload.get("user_id")
        if not user_id:
            raise ValueError("Twitch event has no chatter user")
        text = event.get("message", {}).get("text") if isinstance(event.get("message"), dict) else event.get("message") or event.get("text")
        if text is None:
            raise ValueError("Twitch event has no chat message")
        return InboundEvent(platform="twitch", user_id=str(user_id), reply_to_id=event.get("message_id"), content_type="text", text=text)
    def send_reply(self, event: InboundEvent, reply: OutboundReply) -> None:
        headers = {"Authorization": f"Bearer {self.token}", "Client-Id": self.client_id}
        for message in reply.messages:
            body = {"broadcaster_id": self.broadcaster_id, "sender_id": self.sender_id, "message": message.text or message.media_url or ""}
            if event.reply_to_id:
                body["reply_parent_message_id"] = event.reply_to_id
            response = self.session.post("https://api.twitch.tv/helix/chat/messages", headers=headers, json=body, timeout=30); response.raise_for_status()
            result = response.json()
            if not result.get("data") or not result["data"][0].get("is_sent"):
                raise RuntimeError(f"Twitch chat message was not sent: {result}")

File: php/src/twitch.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function twitch_parse_event(array $payload): array
{
    $event=$payload['event']??$payload;$user=$event['chatter_user_id']??$event['user_id']??'';$text=is_array($event['message']??null)?($event['message']['text']??null):($event['message']??$event['text']??null);if($user===''||$text===null)throw new InvalidArgumentException('Twitch event has no chat message');return ['platform'=>'twitch','user_id'=>(string)$user,'content_type'=>'text','text'=>(string)$text,'reply_to_id'=>$event['message_id']??null];
}
function twitch_send_reply(array $event,array $reply,string $token,string $clientId,string $broadcasterId,string $senderId):void
{
    if($token===''||$clientId===''||$broadcasterId===''||$senderId==='')throw new InvalidArgumentException('Twitch credentials are required');foreach(array_slice($reply['messages']??[],0,5)as$message){if(($message['type']??'text')!=='text')throw new InvalidArgumentException('Twitch chat supports text replies only');$body=['broadcaster_id'=>$broadcasterId,'sender_id'=>$senderId,'message'=>$message['text']??''];if(!empty($event['reply_to_id']))$body['reply_parent_message_id']=$event['reply_to_id'];$ch=curl_init('https://api.twitch.tv/helix/chat/messages');curl_setopt_array($ch,[CURLOPT_POST=>true,CURLOPT_RETURNTRANSFER=>true,CURLOPT_HTTPHEADER=>['Authorization: Bearer '.$token,'Client-Id: '.$clientId,'Content-Type: application/json'],CURLOPT_POSTFIELDS=>json_encode($body,JSON_THROW_ON_ERROR)]);$result=curl_exec($ch);$status=curl_getinfo($ch,CURLINFO_RESPONSE_CODE);$bodyResult=is_string($result)?json_decode($result,true):null;if($result===false||$status<200||$status>=300||empty($bodyResult['data'][0]['is_sent']))throw new RuntimeException('Twitch API request failed');curl_close($ch);}
}
